File: scripts/commands/work_items.py
"""
Work Item (Issue) Commands

Commands for managing work items (issues) in projects.
"""

import os
import re
import sys

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
PARENT_DIR = os.path.dirname(SCRIPT_DIR)
if PARENT_DIR not in sys.path:
    sys.path.insert(0, PARENT_DIR)

from html import escape as html_escape
from urllib.parse import urlencode

from plane_api import api, get_workspace, _validate_id
from plane_output import format_output, green, yellow, PRIORITY_LABELS


def _markdown_inline(text):
    """Convert common inline Markdown to safe HTML.

    This deliberately handles a small, predictable Markdown subset. The
    explicit HTML option remains available for content that needs complete
    control over Plane editor markup.
    """
    escaped = html_escape(text, quote=True)
    protected = []

    def protect(value):
        token = f"\x00{len(protected)}\x00"
        protected.append(value)
        return token

    escaped = re.sub(
        r"`([^`\n]+)`",
        lambda match: protect(f"<code>{match.group(1)}</code>"),
        escaped,
    )

    def replace_link(match):
        label = match.group(1)
        href = match.group(2)
        if not href.startswith(("http://", "https://", "mailto:", "tel:")):
            return match.group(0)
        return protect(f'<a href="{href}">{label}</a>')

    escaped = re.sub(
        r"\[([^\]]+)\]\(((?:https?://|mailto:|tel:)[^\s)]+)\)",
        replace_link,
        escaped,
    )
    escaped = re.sub(
        r"\*\*(.+?)\*\*|__(.+?)__",
        lambda match: f"<strong>{match.group(1) or match.group(2)}</strong>",
        escaped,
    )
    escaped = re.sub(r"~~(.+?)~~", r"<s>\1</s>", escaped)
    escaped = re.sub(r"(?<!\*)\*([^*\n]+)\*(?!\*)", r"<em>\1</em>", escaped)
    escaped = re.sub(r"(?<!\w)_([^_\n]+)_(?!\w)", r"<em>\1</em>", escaped)

    for index, value in enumerate(protected):
        escaped = escaped.replace(f"\x00{index}\x00", value)
    return escaped


def _markdown_complexity_warnings(markdown):
    """Return warnings for Markdown constructs outside the simple converter."""
    warnings = []
    checks = [
        (r"(?m)^\s*\|.*\|\s*$", "tables may not convert correctly"),
        (r"(?m)^\s{2,}[-+*]\s+", "nested lists may not convert correctly"),
        (r"(?m)^\s*[-+*]\s+\[[ xX]\]\s+", "task lists may not convert correctly"),
        (r"!\[[^\]]*\]\(", "images should be added through the Plane editor"),
        (r"\[\^[^\]]+\]", "footnotes are not supported by the simple converter"),
        (r"</?[a-zA-Z][^>]*>", "raw HTML is escaped; use --description-html for HTML"),
    ]
    for pattern, message in checks:
        if re.search(pattern, markdown):
            warnings.append(message)
    return warnings


def _markdown_to_html(markdown):
    """Convert common block-level Markdown to Plane-compatible HTML."""
    lines = markdown.replace("\r\n", "\n").replace("\r", "\n").split("\n")
    blocks = []
    paragraph_lines = []
    list_tag = None
    list_items = []
    quote_lines = []
    code_lines = None
    code_language = ""

    def flush_paragraph():
        if paragraph_lines:
            content = " ".join(line.strip() for line in paragraph_lines)
            blocks.append(f"<p>{_markdown_inline(content)}</p>")
            paragraph_lines.clear()

    def flush_list():
        nonlocal list_tag
        if list_tag:
            blocks.append(f"<{list_tag}>" + "".join(list_items) + f"</{list_tag}>")
            list_tag = None
            list_items.clear()

    def flush_quote():
        if quote_lines:
            content = " ".join(line.strip() for line in quote_lines)
            blocks.append(f"<blockquote><p>{_markdown_inline(content)}</p></blockquote>")
            quote_lines.clear()

    def flush_open_blocks():
        flush_paragraph()
        flush_list()
        flush_quote()

    for line in lines:
        fence = re.match(r"^\s{0,3}(`{3,}|~{3,})([\w+-]*)\s*$", line)
        if code_lines is not None:
            if fence:
                language_attr = (
                    f' class="language-{html_escape(code_language, quote=True)}"' if code_language else ""
                )
                blocks.append(
                    f"<pre><code{language_attr}>{html_escape(chr(10).join(code_lines), quote=False)}</code></pre>"
                )
                code_lines = None
                code_language = ""
            else:
                code_lines.append(line)
            continue

        if fence:
            flush_open_blocks()
            code_lines = []
            code_language = fence.group(2)
            continue

        if not line.strip():
            flush_open_blocks()
            continue

        heading = re.match(r"^\s{0,3}(#{1,6})\s+(.+?)\s*#*\s*$", line)
        if heading:
            flush_open_blocks()
            level = len(heading.group(1))
            blocks.append(f"<h{level}>{_markdown_inline(heading.group(2))}</h{level}>")
            continue

        if re.match(r"^\s{0,3}((\*\s*){3,}|(-\s*){3,}|(_\s*){3,})$", line):
            flush_open_blocks()
            blocks.append("<hr>")
            continue

        quote = re.match(r"^\s{0,3}>\s?(.*)$", line)
        if quote:
            flush_paragraph()
            flush_list()
            quote_lines.append(quote.group(1))
            continue
        flush_quote()

        unordered = re.match(r"^\s*[-+*]\s+(.*)$", line)
        ordered = re.match(r"^\s*\d+[.)]\s+(.*)$", line)
        if unordered or ordered:
            flush_paragraph()
            new_list_tag = "ul" if unordered else "ol"
            if list_tag and list_tag != new_list_tag:
                flush_list()
            list_tag = list_tag or new_list_tag
            item = unordered.group(1) if unordered else ordered.group(1)
            list_items.append(f"<li>{_markdown_inline(item)}</li>")
            continue
        flush_list()

        paragraph_lines.append(line)

    if code_lines is not None:
        language_attr = f' class="language-{html_escape(code_language, quote=True)}"' if code_language else ""
        blocks.append(
            f"<pre><code{language_attr}>{html_escape(chr(10).join(code_lines), quote=False)}</code></pre>"
        )
    else:
        flush_open_blocks()

    return "".join(blocks)


def _get_description_html(args):
    """Return the description payload while preserving explicit Plane HTML.

    ``--description`` accepts plain text and common Markdown. Complex Markdown
    may not convert perfectly, so ``--description-html`` remains the explicit
    path for Plane editor markup such as ``data-text-color`` attributes.
    """
    description_html = getattr(args, "description_html", None)
    description = getattr(args, "description", None)
    if description_html is not None:
        return description_html
    if description is not None:
        for warning in _markdown_complexity_warnings(description):
            print(yellow(f"Warning: {warning}; use --description-html for full control."), file=sys.stderr)
        return _markdown_to_html(description)
    return None


def cmd_issues_list(args):
    """List work items in a project, with optional filters and pagination."""
    _validate_id(args.project, "project ID")
    endpoint = f"/workspaces/{get_workspace()}/projects/{args.project}/work-items/"
    params = {}

    # Filters
    if args.state:
        params["state"] = args.state
    if args.priority:
        params["priority"] = args.priority
    if args.assignee:
        params["assignees"] = args.assignee

    # Pagination
    if hasattr(args, 'cursor') and args.cursor:
        params["cursor"] = args.cursor
    if hasattr(args, 'per_page') and args.per_page:
        params["per_page"] = args.per_page

    # Field selection and expansion
    if hasattr(args, 'fields') and args.fields:
        params["fields"] = args.fields
    if hasattr(args, 'expand') and args.expand:
        params["expand"] = args.expand

    # Ordering
    if hasattr(args, 'order_by') and args.order_by:
        params["order_by"] = args.order_by

    if params:
        endpoint += "?" + urlencode(params)
    data = api("GET", endpoint)
    format_output(data, args.format)


def cmd_issues_get(args):
    """Get full details of a single work item."""
    _validate_id(args.project, "project ID")
    _validate_id(args.issue, "issue ID")
    data = api("GET", f"/workspaces/{get_workspace()}/projects/{args.project}/work-items/{args.issue}/")
    format_output(data, args.format)


def cmd_issues_create(args):
    """Create a new work item in a project."""
    _validate_id(args.project, "project ID")
    payload = {"name": args.name}
    description_html = _get_description_html(args)
    if description_html is not None:
        payload["description_html"] = description_html
    if args.priority:
        payload["priority"] = PRIORITY_LABELS.get(args.priority.lower(), 3)
    if args.state:
        payload["state"] = args.state
    if args.assignee:
        payload["assignees"] = [args.assignee]
    if args.label:
        payload["labels"] = [args.label]

    data = api("POST", f"/workspaces/{get_workspace()}/projects/{args.project}/work-items/", payload)
    print(green("✓ Work item created"))
    format_output(data, args.format)


def cmd_issues_update(args):
    """Update fields on an existing work item."""
    _validate_id(args.project, "project ID")
    _validate_id(args.issue, "issue ID")
    payload = {}
    if args.name:
        payload["name"] = args.name
    description_html = _get_description_html(args)
    if description_html is not None:
        payload["description_html"] = description_html
    if args.priority:
        payload["priority"] = PRIORITY_LABELS.get(args.priority.lower(), 3)
    if args.state:
        payload["state"] = args.state

    if not payload:
        print(yellow("Nothing to update — pass at least one field."), file=sys.stderr)
        sys.exit(1)

    data = api("PATCH", f"/workspaces/{get_workspace()}/projects/{args.project}/work-items/{args.issue}/", payload)
    print(green("✓ Work item updated"))
    format_output(data, args.format)


def cmd_issues_assign(args):
    """Assign a work item to one or more members."""
    _validate_id(args.project, "project ID")
    _validate_id(args.issue, "issue ID")
    for assignee in args.assignees:
        _validate_id(assignee, "assignee ID")
    payload = {"assignees": args.assignees}
    data = api("PATCH", f"/workspaces/{get_workspace()}/projects/{args.project}/work-items/{args.issue}/", payload)
    print(green(f"✓ Assigned to {len(args.assignees)} member(s)"))
    format_output(data, args.format)


def cmd_issues_delete(args):
    """Delete a work item (irreversible)."""
    _validate_id(args.project, "project ID")
    _validate_id(args.issue, "issue ID")
    api("DELETE", f"/workspaces/{get_workspace()}/projects/{args.project}/work-items/{args.issue}/")
    print(green(f"✓ Deleted work item {args.issue}"))


def cmd_issues_search(args):
    """Search work items across the entire workspace."""
    params = urlencode({"search": args.query, "type": "work_item"})
    endpoint = f"/workspaces/{get_workspace()}/search/?{params}"
    data = api("GET", endpoint)
    format_output(data, args.format)


def cmd_issues_get_short(args):
    """Get work item details by short ID (e.g., PROJ-123)."""
    _validate_id(args.issue_short, "issue short ID")
    data = api("GET", f"/workspaces/{get_workspace()}/work-items/{args.issue_short}/")
    format_output(data, args.format)

File: internal/commands/work_items.go
package commands

import (
	"errors"
	"fmt"
	"net/url"

	"github.com/HonLuk/my-plane/internal/markdown"
)

func (r *runner) runIssues(args []string) error {
	if len(args) == 0 || args[0] == "--help" || args[0] == "-h" {
		r.output.Println(`Usage: plane issues <list|get|create|update|assign|delete|search|get-short> [options]

List, create, update, assign, delete, and search work items.`)
		return nil
	}
	switch args[0] {
	case "list":
		return r.issuesList(args[1:])
	case "get":
		return r.issuesGet(args[1:])
	case "create":
		return r.issuesCreate(args[1:])
	case "update":
		return r.issuesUpdate(args[1:])
	case "assign":
		return r.issuesAssign(args[1:])
	case "delete":
		return r.issuesDelete(args[1:])
	case "search":
		return r.issuesSearch(args[1:])
	case "get-short":
		return r.issuesGetShort(args[1:])
	default:
		return fmt.Errorf("unknown issues subcommand %q", args[0])
	}
}

func addProjectAliases(set interface {
	String(string, string, string) *string
	StringVar(*string, string, string, string)
}) *string {
	project := set.String("project", "", "Project ID (UUID)")
	set.StringVar(project, "p", "", "Project ID (UUID)")
	return project
}

func (r *runner) issuesList(args []string) error {
	set := r.newFlagSet("issues list", "Usage: plane issues list --project PROJECT_ID [options]\n\nShow work items with optional filters for state, priority, or assignee.")
	options := addCommon(set, true)
	project := addProjectAliases(set)
	state := set.String("state", "", "Filter by state ID")
	priority := set.String("priority", "", "Filter by priority level")
	assignee := set.String("assignee", "", "Filter by assignee user ID")
	orderBy := set.String("order-by", "", "Sort field (e.g., '-created_at', 'priority')")
	help, err := parseFlags(set, args)
	if help || err != nil {
		return err
	}
	if *project == "" {
		return errors.New("usage: plane issues list --project PROJECT_ID")
	}
	if err := validatePriority(*priority); err != nil {
		return err
	}
	if err := validateFormat(options.format); err != nil {
		return err
	}
	if err := validateIDs(
		struct{ value, name string }{*project, "project ID"},
		struct{ value, name string }{*state, "state ID"},
		struct{ value, name string }{*assignee, "assignee ID"},
	); err != nil {
		return err
	}
	client, err := r.client()
	if err != nil {
		return err
	}
	params := url.Values{}
	if *state != "" {
		params.Set("state", *state)
	}
	if *priority != "" {
		params.Set("priority", *priority)
	}
	if *assignee != "" {
		params.Set("assignees", *assignee)
	}
	addPagination(params, options)
	if options.fields != "" {
		params.Set("fields", options.fields)
	}
	if options.expand != "" {
		params.Set("expand", options.expand)
	}
	if *orderBy != "" {
		params.Set("order_by", *orderBy)
	}
	data, err := r.request(client, "GET", endpoint("/workspaces/"+client.Workspace+"/projects/"+*project+"/work-items/", params), nil)
	if err != nil {
		return err
	}
	r.output.Format(data, options.format, true)
	return nil
}

func (r *runner) issuesGet(args []string) error {
	set := r.newFlagSet("issues get", "Usage: plane issues get --project PROJECT_ID ISSUE_ID [options]\n\nView full details of a specific work item by UUID.")
	options := addCommon(set, false)
	project := addProjectAliases(set)
	help, err := parseFlags(set, args)
	if help || err != nil {
		return err
	}
	if *project == "" || len(set.Args()) != 1 {
		return errors.New("usage: plane issues get --project PROJECT_ID ISSUE_ID")
	}
	issue := set.Args()[0]
	if err := validateFormat(options.format); err != nil {
		return err
	}
	if err := validateIDs(struct{ value, name string }{*project, "project ID"}, struct{ value, name string }{issue, "issue ID"}); err != nil {
		return err
	}
	client, err := r.client()
	if err != nil {
		return err
	}
	data, err := r.request(client, "GET", "/workspaces/"+client.Workspace+"/projects/"+*project+"/work-items/"+issue+"/", nil)
	if err != nil {
		return err
	}
	r.output.Format(data, options.format, true)
	return nil
}

func (r *runner) issuesCreate(args []string) error {
	set := r.newFlagSet("issues create", "Usage: plane issues create --project PROJECT_ID --name NAME [options]\n\nCreate a work item with title and optional fields.")
	options := addCommon(set, false)
	project := addProjectAliases(set)
	name := set.String("name", "", "Work item title")
	description := set.String("description", "", "Plain text or simple Markdown")
	descriptionHTML := set.String("description-html", "", "Plane editor HTML")
	priority := set.String("priority", "", "Priority level")
	state := set.String("state", "", "State ID")
	assignee := set.String("assignee", "", "Assignee user ID")
	label := set.String("label", "", "Label ID")
	help, err := parseFlags(set, args)
	if help || err != nil {
		return err
	}
	if *project == "" || *name == "" {
		return errors.New("usage: plane issues create --project PROJECT_ID --name NAME")
	}
	if flagWasSet(set, "description") && flagWasSet(set, "description-html") {
		return errors.New("--description and --description-html are mutually exclusive")
	}
	if err := validatePriority(*priority); err != nil {
		return err
	}
	if err := validateFormat(options.format); err != nil {
		return err
	}
	if err := validateIDs(
		struct{ value, name string }{*project, "project ID"},
		struct{ value, name string }{*state, "state ID"},
		struct{ value, name string }{*assignee, "assignee ID"},
		struct{ value, name string }{*label, "label ID"},
	); err != nil {
		return err
	}
	descriptionValue, hasDescription, err := r.descriptionHTML(*description, *descriptionHTML)
	if err != nil {
		return err
	}
	client, err := r.client()
	if err != nil {
		return err
	}
	payload := map[string]any{"name": *name}
	if hasDescription {
		payload["description_html"] = descriptionValue
	}
	if *priority != "" {
		payload["priority"] = *priority
	}
	if *state != "" {
		payload["state"] = *state
	}
	if *assignee != "" {
		payload["assignees"] = []string{*assignee}
	}
	if *label != "" {
		payload["labels"] = []string{*label}
	}
	data, err := r.request(client, "POST", "/workspaces/"+client.Workspace+"/projects/"+*project+"/work-items/", payload)
	if err != nil {
		return err
	}
	r.output.Println(r.output.Green("✓ Work item created"))
	r.output.Format(data, options.format, true)
	return nil
}

func (r *runner) issuesUpdate(args []string) error {
	set := r.newFlagSet("issues update", "Usage: plane issues update --project PROJECT_ID ISSUE_ID [options]\n\nModify title, description, priority, or state of a work item.")
	options := addCommon(set, false)
	project := addProjectAliases(set)
	name := set.String("name", "", "New title")
	description := set.String("description", "", "New plain text or simple Markdown")
	descriptionHTML := set.String("description-html", "", "New Plane editor HTML")
	priority := set.String("priority", "", "New priority level")
	state := set.String("state", "", "New state ID")
	help, err := parseFlags(set, args)
	if help || err != nil {
		return err
	}
	if *project == "" || len(set.Args()) != 1 {
		return errors.New("usage: plane issues update --project PROJECT_ID ISSUE_ID")
	}
	if flagWasSet(set, "description") && flagWasSet(set, "description-html") {
		return errors.New("--description and --description-html are mutually exclusive")
	}
	if err := validatePriority(*priority); err != nil {
		return err
	}
	if err := validateFormat(options.format); err != nil {
		return err
	}
	issue := set.Args()[0]
	if err := validateIDs(struct{ value, name string }{*project, "project ID"}, struct{ value, name string }{issue, "issue ID"}, struct{ value, name string }{*state, "state ID"}); err != nil {
		return err
	}
	descriptionValue, hasDescription, err := r.descriptionHTML(*description, *descriptionHTML)
	if err != nil {
		return err
	}
	payload := map[string]any{}
	if *name != "" {
		payload["name"] = *name
	}
	if hasDescription {
		payload["description_html"] = descriptionValue
	}
	if *priority != "" {
		payload["priority"] = *priority
	}
	if *state != "" {
		payload["state"] = *state
	}
	if len(payload) == 0 {
		return errors.New("Nothing to update — pass at least one field.")
	}
	client, err := r.client()
	if err != nil {
		return err
	}
	data, err := r.request(client, "PATCH", "/workspaces/"+client.Workspace+"/projects/"+*project+"/work-items/"+issue+"/", payload)
	if err != nil {
		return err
	}
	r.output.Println(r.output.Green("✓ Work item updated"))
	r.output.Format(data, options.format, true)
	return nil
}

func (r *runner) issuesAssign(args []string) error {
	set := r.newFlagSet("issues assign", "Usage: plane issues assign --project PROJECT_ID ISSUE_ID USER_ID [USER_ID...]\n\nAssign a work item to one or more workspace members.")
	options := addCommon(set, false)
	project := addProjectAliases(set)
	help, err := parseFlags(set, args)
	if help || err != nil {
		return err
	}
	if *project == "" || len(set.Args()) < 2 {
		return errors.New("usage: plane issues assign --project PROJECT_ID ISSUE_ID USER_ID [USER_ID...]")
	}
	issue := set.Args()[0]
	assignees := set.Args()[1:]
	if err := validateFormat(options.format); err != nil {
		return err
	}
	if err := validateIDs(struct{ value, name string }{*project, "project ID"}, struct{ value, name string }{issue, "issue ID"}); err != nil {
		return err
	}
	for _, assignee := range assignees {
		if err := validateIDs(struct{ value, name string }{assignee, "assignee ID"}); err != nil {
			return err
		}
	}
	client, err := r.client()
	if err != nil {
		return err
	}
	payload := map[string]any{"assignees": assignees}
	data, err := r.request(client, "PATCH", "/workspaces/"+client.Workspace+"/projects/"+*project+"/work-items/"+issue+"/", payload)
	if err != nil {
		return err
	}
	r.output.Println(r.output.Green(fmt.Sprintf("✓ Assigned to %d member(s)", len(assignees))))
	r.output.Format(data, options.format, true)
	return nil
}

func (r *runner) issuesDelete(args []string) error {
	set := r.newFlagSet("issues delete", "Usage: plane issues delete --project PROJECT_ID ISSUE_ID [options]\n\nPermanently delete a work item (irreversible).")
	options := addCommon(set, false)
	project := addProjectAliases(set)
	help, err := parseFlags(set, args)
	if help || err != nil {
		return err
	}
	if *project == "" || len(set.Args()) != 1 {
		return errors.New("usage: plane issues delete --project PROJECT_ID ISSUE_ID")
	}
	if err := validateFormat(options.format); err != nil {
		return err
	}
	issue := set.Args()[0]
	if err := validateIDs(struct{ value, name string }{*project, "project ID"}, struct{ value, name string }{issue, "issue ID"}); err != nil {
		return err
	}
	client, err := r.client()
	if err != nil {
		return err
	}
	if _, err := r.request(client, "DELETE", "/workspaces/"+client.Workspace+"/projects/"+*project+"/work-items/"+issue+"/", nil); err != nil {
		return err
	}
	r.output.Println(r.output.Green("✓ Deleted work item " + issue))
	return nil
}

func (r *runner) issuesSearch(args []string) error {
	set := r.newFlagSet("issues search", "Usage: plane issues search QUERY [options]\n\nSearch work items by text across all projects in the workspace.")
	options := addCommon(set, false)
	help, err := parseFlags(set, args)
	if help || err != nil {
		return err
	}
	if err := requirePositionals(set.Args(), 1, "plane issues search QUERY"); err != nil {
		return err
	}
	if err := validateFormat(options.format); err != nil {
		return err
	}
	client, err := r.client()
	if err != nil {
		return err
	}
	params := url.Values{"search": []string{set.Args()[0]}, "type": []string{"work_item"}}
	data, err := r.request(client, "GET", endpoint("/workspaces/"+client.Workspace+"/search/", params), nil)
	if err != nil {
		return err
	}
	r.output.Format(data, options.format, true)
	return nil
}

func (r *runner) issuesGetShort(args []string) error {
	set := r.newFlagSet("issues get-short", "Usage: plane issues get-short PROJECT-SEQ [options]\n\nQuick access to a work item using its short identifier format.")
	options := addCommon(set, false)
	help, err := parseFlags(set, args)
	if help || err != nil {
		return err
	}
	if err := requirePositionals(set.Args(), 1, "plane issues get-short PROJECT-SEQ"); err != nil {
		return err
	}
	issueShort := set.Args()[0]
	if err := validateFormat(options.format); err != nil {
		return err
	}
	if err := validateIDs(struct{ value, name string }{issueShort, "issue short ID"}); err != nil {
		return err
	}
	client, err := r.client()
	if err != nil {
		return err
	}
	data, err := r.request(client, "GET", "/workspaces/"+client.Workspace+"/work-items/"+issueShort+"/", nil)
	if err != nil {
		return err
	}
	r.output.Format(data, options.format, true)
	return nil
}

func validatePriority(value string) error {
	if value == "" {
		return nil
	}
	switch value {
	case "urgent", "high", "medium", "low", "none":
		return nil
	default:
		return fmt.Errorf("invalid priority %q: choose urgent, high, medium, low, or none", value)
	}
}

func (r *runner) descriptionHTML(description, explicitHTML string) (string, bool, error) {
	if description != "" && explicitHTML != "" {
		return "", false, errors.New("--description and --description-html are mutually exclusive")
	}
	if explicitHTML != "" {
		return explicitHTML, true, nil
	}
	if description == "" {
		return "", false, nil
	}
	for _, warning := range markdown.Warnings(description) {
		r.output.Errorln(r.output.Yellow("Warning: " + warning + "; use --description-html for full control."))
	}
	return markdown.ToHTML(description), true, nil
}
